### PR TITLE
EIP-2565 #2328

### PR DESCRIPTION
**Pull request by quilir**
New implementation of ModExp inside Evm.Precompiles

This change requires the closing of https://github.com/NethermindEth/Math.Gmp.Native/pull/3
Link to benchmarks: https://pastebin.com/2RUuUHv5

### CODE REVIEW

File: src/Nethermind/Nethermind.Evm/Precompiles/ModExpPrecompile2565.cs [hunk 1]
```diff
@@ -91,6 +104,53 @@ public long DataGasCost(byte[] inputData, IReleaseSpec releaseSpec)
...
+            
+            
+            using void_ptr data = gmp_lib.allocate((size_t) modulusLength);
+            ptr<size_t> countp = new(0);
+            gmp_lib.mpz_export(data, countp, 1, 1, 1, 0, powmResult);
+            int count = (int) countp.Value;
+
+
+            byte[] result = new byte[modulusLength];
+            Marshal.Copy(data.ToIntPtr(), result, modulusLength - count, count);
```
Comment (tkstanczak) [inline src/Nethermind/Nethermind.Evm/Precompiles/ModExpPrecompile2565.cs:142]: don't we leak data without deallocating here?

Comment (tkstanczak) [reply src/Nethermind/Nethermind.Evm/Precompiles/ModExpPrecompile2565.cs:142]: shall we deallocate in finally?

File: src/Nethermind/Nethermind.Precompiles.Benchmark/Program.cs [hunk 2]
```diff
@@ -35,9 +35,9 @@ public static void Main(string[] args)
             // BenchmarkRunner.Run<Bn256MulBenchmark>();
             // BenchmarkRunner.Run<Bn256PairingBenchmark>();
             BenchmarkRunner.Run<ModExp2565Benchmark>();
```
Comment (tkstanczak) [inline src/Nethermind/Nethermind.Precompiles.Benchmark/Program.cs:37]: I cannot find the changes / results of the eip2565 benchmark

File: src/Nethermind/Nethermind.Evm/Precompiles/ModExpPrecompile2565.cs (hunk 1 above)
Comment (LukaszRozmej) [reply src/Nethermind/Nethermind.Evm/Precompiles/ModExpPrecompile2565.cs:142]: using void_ptr data = gmp_lib.allocate((size_t) modulusLength);

we made data  disposable, then its deallocated

**Review by Az111z — APPROVED** [review]

**Comment by LukaszRozmej** [comment]
My latest run of benchmarks:

BenchmarkDotNet=v0.12.1, OS=Windows 10.0.19042
AMD Ryzen 9 5900X, 1 CPU, 24 logical and 12 physical cores
.NET Core SDK=5.0.201
  [Host]        : .NET Core 5.0.4 (CoreCLR 5.0.421.11614, CoreFX 5.0.421.11614), X64 RyuJIT
  .NET Core 5.0 : .NET Core 5.0.4 (CoreCLR 5.0.421.11614, CoreFX 5.0.421.11614), X64 RyuJIT

Job=.NET Core 5.0  Runtime=.NET Core 5.0  

|   Method |                Input |             Mean |           Error |          StdDev | Ratio | RatioSD |
|--------- |--------------------- |-----------------:|----------------:|----------------:|------:|--------:|
|   BigInt |         eip_example1 |     112,806.4 ns |     2,240.05 ns |     4,626.09 ns | 12.50 |    0.54 |
| Baseline |         eip_example1 |       9,416.1 ns |        95.14 ns |        84.34 ns |  1.00 |    0.00 |
|          |                      |                  |                 |                 |       |         |
|   BigInt |         eip_example2 |       6,705.4 ns |       122.03 ns |       130.58 ns |  3.32 |    0.06 |
| Baseline |         eip_example2 |       2,027.6 ns |        14.97 ns |        14.00 ns |  1.00 |    0.00 |
|          |                      |                  |                 |                 |       |         |
|   BigInt | nagyd(...)10001 [21] |       8,770.6 ns |        64.34 ns |        60.18 ns |  2.34 |    0.02 |
| Baseline | nagyd(...)10001 [21] |       3,745.8 ns |        19.29 ns |        18.04 ns |  1.00 |    0.00 |
|          |                      |                  |                 |                 |       |         |
|   BigInt |      nagydani-1-qube |       1,413.1 ns |         8.82 ns |         7.82 ns |  0.51 |    0.01 |
| Baseline |      nagydani-1-qube |       2,747.4 ns |        23.43 ns |        20.77 ns |  1.00 |    0.00 |
|          |                      |                  |                 |                 |       |         |
|   BigInt |    nagydani-1-square |         891.0 ns |         5.85 ns |         5.19 ns |  0.32 |    0.00 |
| Baseline |    nagydani-1-square |       2,760.8 ns |        11.24 ns |         9.38 ns |  1.00 |    0.00 |
|          |                      |                  |                 |                 |       |         |
|   BigInt | nagyd(...)10001 [21] |      36,026.2 ns |       123.23 ns |       115.27 ns |  4.81 |    0.03 |
| Baseline | nagyd(...)10001 [21] |       7,482.4 ns |        24.68 ns |        23.09 ns |  1.00 |    0.00 |
|          |                      |                  |                 |                 |       |         |
|   BigInt |      nagydani-2-qube |       6,239.7 ns |        42.29 ns |        35.31 ns |  1.56 |    0.01 |
| Baseline |      nagydani-2-qube |       3,994.6 ns |        20.86 ns |        19.51 ns |  1.00 |    0.00 |
|          |                      |                  |                 |                 |       |         |
|   BigInt |    nagydani-2-square |       4,017.0 ns |        12.32 ns |        11.52 ns |  1.10 |    0.01 |
| Baseline |    nagydani-2-square |       3,667.4 ns |        29.04 ns |        25.75 ns |  1.00 |    0.00 |
|          |                      |                  |                 |                 |       |         |
|   BigInt | nagyd(...)10001 [21] |     113,357.8 ns |       607.82 ns |       568.55 ns |  5.48 |    0.04 |
| Baseline | nagyd(...)10001 [21] |      20,685.8 ns |        95.59 ns |        89.42 ns |  1.00 |    0.00 |
|          |                      |                  |                 |                 |       |         |
|   BigInt |      nagydani-3-qube |      18,783.3 ns |        89.68 ns |        83.88 ns |  2.48 |    0.02 |
| Baseline |      nagydani-3-qube |       7,582.5 ns |        35.71 ns |        33.40 ns |  1.00 |    0.00 |
|          |                      |                  |                 |                 |       |         |
|   BigInt |    nagydani-3-square |      12,088.2 ns |        67.61 ns |        52.79 ns |  1.83 |    0.01 |
| Baseline |    nagydani-3-square |       6,596.1 ns |        19.13 ns |        17.89 ns |  1.00 |    0.00 |
|          |                      |                  |                 |                 |       |         |
|   BigInt | nagyd(...)10001 [21] |     370,721.0 ns |     1,040.65 ns |       922.50 ns |  5.24 |    0.02 |
| Baseline | nagyd(...)10001 [21] |      70,734.2 ns |       251.72 ns |       223.14 ns |  1.00 |    0.00 |
|          |                      |                  |                 |                 |       |         |
|   BigInt |      nagydani-4-qube |      73,761.0 ns |       420.50 ns |       393.33 ns |  3.75 |    0.02 |
| Baseline |      nagydani-4-qube |      19,636.0 ns |        73.96 ns |        61.76 ns |  1.00 |    0.00 |
|          |                      |                  |                 |                 |       |         |
|   BigInt |    nagydani-4-square |      51,669.3 ns |       208.13 ns |       194.69 ns |  3.20 |    0.02 |
| Baseline |    nagydani-4-square |      16,134.9 ns |       110.41 ns |       103.27 ns |  1.00 |    0.00 |
|          |                      |                  |                 |                 |       |         |
|   BigInt | nagyd(...)10001 [21] |   1,282,100.1 ns |     5,931.34 ns |     5,257.97 ns |  6.91 |    0.03 |
| Baseline | nagyd(...)10001 [21] |     185,509.9 ns |       450.22 ns |       421.13 ns |  1.00 |    0.00 |
|          |                      |                  |                 |                 |       |         |
|   BigInt |      nagydani-5-qube |     352,078.5 ns |     1,240.67 ns |     1,160.52 ns |  6.78 |    0.03 |
| Baseline |      nagydani-5-qube |      51,935.4 ns |       193.18 ns |       180.70 ns |  1.00 |    0.00 |
|          |                      |                  |                 |                 |       |         |
|   BigInt |    nagydani-5-square |     284,196.1 ns |     1,480.40 ns |     1,384.76 ns |  6.78 |    0.04 |
| Baseline |    nagydani-5-square |      41,925.5 ns |       201.77 ns |       188.74 ns |  1.00 |    0.00 |
|          |                      |                  |                 |                 |       |         |
|   BigInt |             random-1 |     970,079.0 ns |     5,710.54 ns |     5,341.65 ns | 22.93 |    0.11 |
| Baseline |             random-1 |      42,333.9 ns |       119.12 ns |        99.47 ns |  1.00 |    0.00 |
|          |                      |                  |                 |                 |       |         |
|   BigInt |             random-2 |  19,799,304.0 ns |   115,563.46 ns |   102,444.00 ns |  9.95 |    0.06 |
| Baseline |             random-2 |   1,990,136.5 ns |    10,626.63 ns |     9,420.23 ns |  1.00 |    0.00 |
|          |                      |                  |                 |                 |       |         |
|   BigInt |             random-3 | 362,474,553.8 ns | 2,377,013.12 ns | 1,984,913.61 ns |  9.30 |    0.15 |
| Baseline |             random-3 |  38,988,251.4 ns |   588,035.88 ns |   550,049.12 ns |  1.00 |    0.00 |

**Review by LukaszRozmej — APPROVED** [review]

**Pull request merged**